### ISSUE TITLE
Support CSS font size keywords

### ISSUE THREAD

**Issue by mhsmith** [beeware/toga#1814]
From https://github.com/beeware/toga/pull/1761#issuecomment-1468691931:

> I've added a `font_size` example app for manual testing. Not all the widgets implement font size yet (that can wait until future PRs), but those that do are mostly consistent across the 5 platforms, and proportionate to the default font size.
> 
> The exception is Windows, where [the default font size is "8 points"](https://learn.microsoft.com/en-us/dotnet/api/system.drawing.systemfonts.defaultfont). However, I think it's actually bigger than 8/72 inches on most screens, because it was fixed at a time when Windows assumed 96 dpi but the typical monitor was less dense than that (see [this article from 1999](https://tidbits.com/1999/02/15/why-windows-web-pages-have-tiny-text/)). And since Windows is so serious about backward compatibility, we're still stuck with it.
> 
> So Toga widgets with fixed font sizes currently appear significantly larger on Windows than on the other platforms. I don't think there's anything we can do about that now, but in the future we could deal with it by implementing the CSS font size keywords `small`, `medium`, `large`, etc.

**Comment by kiljoy001** [comment]
How can I get started on implementing this issue?

**Comment by freakboy3742** [comment]
There's two pieces of work required.

The first is a modification to the Pack definition (core/src/toga/style/pack) to change `FONT_SIZE_CHOICES` to allow the new keywords for various font sizes. Using the CSS size definitions would be a good start; it might be advisable to add those constants to the underlying Travertino library, rather than Toga.

Next, the font definition on each platform needs to be modified to allow for these keywords when specifying font sizes. Again, you can follow CSS here; whatever "normal" font size is on each platform, each bump up/down should be about 20% larger/smaller. This will vary on each platform; on Cocoa, "normal" is 12pt; but on Android, "normal" is widget-specific, and specified in `sp` (14sp on button, for example).

You don't need to implement it on every platform, but it needs to not be *explosive* on every platform - so at the very least you'll need to add warnings on all platforms to flag that named font sizes will be ignored.

**Comment by freakboy3742** [comment]
I've added the required constants in beeware/travertino#43. We're overdue for a travertino release; once those constants are in place, I'll cut a release.

**Comment by mhsmith** [comment]
> whatever "normal" font size is on each platform, each bump up/down should be about 20% larger/smaller. This will vary on each platform; on Cocoa, "normal" is 12pt; but on Android, "normal" is widget-specific, and specified in `sp` (14sp on button, for example).

On Android, the only widget we've encountered so far with a different default font size is the TextInput, which is larger than the others. But I think it's better for "normal" to be absolute, and defined by whatever is the most common font size across the most common widgets. For example a file manager app may implement file renames by changing the filename from a Label to a TextInput, while keeping the font size the same to maintain visual continuity.

If you want to use the widget's default font size as a baseline,  that's what "larger" and "smaller" are for. These were also added in https://github.com/beeware/travertino/pull/43, but they don't necessarily need to be implemented at the same time as the absolute keywords.

**Comment by dkarvan8** [comment]
Hi, I want to work on this issue. Can you assign it to me?

**Comment by mhsmith** [comment]
We don't use the GitHub assignment feature – posting a comment is enough.

You've posted this comment on 5 issues today, but I recommend you only work on one at a time until you're more familiar with the project.

**Comment by sasoder** [comment]
Hi! I’m interested in working on this issue as part of a software engineering course. Is it still available? If so, I’d love to tackle it!

**Comment by mhsmith** [comment]
Yes, it's still available. Travertino is now part of the Toga repository, but apart from that I think the comments above are still valid.

**Comment by sasoder** [comment]
Cool, thanks for the heads up. I'll get started on it!

**Comment by Jacmol** [comment]
Hi, working on this with @sasoder and had a question about how you could go about implementing this on GTK? As I understand it when setting a system default font size, Pango won't apply a value until after the font is used on a widget, and until then it is 0. How would you then go about setting a value that is relative to the default size?

**Comment by freakboy3742** [comment]
Two other approaches that might be worth exploring before committing to a hard-coded value:

* Assign a font size, determine how Pango has interpreted it, and then immediately change the font size to the "final" value. We need to be careful that we don't introduce "double render" artefacts if we take this approach, but if the call is done without releasing control to the event loop, it *might* work.

* GTK's underlying style system is ultimately CSS. I haven't explored if GTK supports any useful "font size" keywords, but if it does, it might be possible to plug that size directly into the CSS rendering by treating "smaller" as a "native representation" of font size.

**Comment by mhsmith** [comment]
> Assign a font size, determine how Pango has interpreted it, and then immediately change the font size to the "final" value. We need to be careful that we don't introduce "double render" artefacts if we take this approach, but if the call is done without releasing control to the event loop, it _might_ work.

And we should only need to do this once – after we know what the default font size is, we can cache it.
